### DISCUSSION TITLE
What does payment status "pending" mean?

### DISCUSSION THREAD

**Discussion by BenSturmfels** [houdiniproject/houdini#1102] [Q&A]
I've put through some test payments through via a Stripe test API key. When I click on a payment in Houdini, I notice that it's listed as "pending". What does "pending" signify here? Is there a background job or similar that talks to Stripe to determine when these payments are settled?

**Comment by wwahammy** [comment]
Ruby `Charge` objects can be in a number of different states in Houdini:

* failed: a charge was attempted but didn't succeed
* pending: Stripe may not have to wait for the charge to settle before finalizing the charge. The money usually isn't in the Stripe account  at that point
* available: The money is available on Stripe
* disbursed: The money is paid out to the nonprofit's bank account (only really useful if using Stripe Connect)


You should run `bin/rails heroku_scheduled_job[update_np_balances]` daily to update that state.

Longer term, we will just handle the webhook from Stripe so there's no need for the job.

There is a background job but it's not documented!

* `bin/rails heroku_scheduled_job[update_np_balances]` will check if the items are still pending. Running it about once a day is ideal.

Ideally, we should handle the webhook from Stripe to simply move charges from pending to available

**Reply by BenSturmfels** [reply]
Thanks very much @wwahammy. Diving into this, I see that the `mark_available_charges` expects a `stripe_account_id`. Can `stripe_account_id` be set via the dashboard in Houdini 2? It appears that a self-hosted non-profit would need to specify `STRIPE_API_KEY`, `STRIPE_API_PUBLIC` environment variables and `stripe_account_id` in the database. Is that right?

**Reply by BenSturmfels** [reply]
Also, would it make sense to extend `crontab.sh` to also run `update_np_balances` as well as the other tasks defined in `scheduled_jobs.rb`?

**Reply by wwahammy** [reply]
> Thanks very much @wwahammy. Diving into this, I see that the `mark_available_charges` expects a `stripe_account_id`. Can `stripe_account_id` be set via the dashboard in Houdini 2? It appears that a self-hosted non-profit would need to specify `STRIPE_API_KEY`, `STRIPE_API_PUBLIC` environment variables and `stripe_account_id` in the database. Is that right?

That does seem to be how it's set up. In cases where there's only a single nonprofit on the instance, that is a pretty awkward setup. I would recommend filing a bug about not having a good way to set the stripe_account_id when nonprofits are created via the command line (which is how they're created in Houdini as of now)

**Reply by wwahammy** [reply]
> Also, would it make sense to extend `crontab.sh` to also run `update_np_balances` as well as the other tasks defined in `scheduled_jobs.rb`?

Yes, good idea!

**Comment by RamtinKosari** [comment]
Hi @BenSturmfels 
In the context of Stripe, when a payment is listed as "pending," it means that the payment is in a state of being processed or awaiting further action. This status typically occurs when the payment is initially created or when it requires additional verification or authorization before it can be considered successful.

During the pending state, Stripe performs various checks and processes to ensure the payment's validity and security. This may involve verifying the payment details, performing risk assessments, and complying with anti-fraud measures. The specific checks and duration of the pending state can vary depending on the payment method, transaction amount, and other factors.

Once Stripe completes the necessary processing and verification, the payment status can change to "succeeded" if the payment is successful or "failed" if it encounters an issue. The transition from "pending" to the final state is typically handled by Stripe's backend processes and does not require manual intervention.

In Houdini, the "pending" status indicates that the payment is currently being processed by Stripe. Houdini likely communicates with Stripe's API to retrieve the payment status and displays it as "pending" until a final status update is received.

If you want to get real-time updates on payment status changes, you can configure webhooks in your Stripe account. Webhooks allow Stripe to send notifications to your application or server when significant events, such as payment status changes, occur. By handling these webhooks, you can update your application's UI or trigger specific actions based on the changing payment status.

Remember that the behavior and terminology may vary depending on the specific integration or application you're using with Stripe. It's always a good idea to consult the documentation or support resources provided by Houdini or Stripe for more detailed information on payment statuses and integration specifics.
